### PR TITLE
RD-6709 Add proper authentication and authorization to snapshot endpoints

### PR DESCRIPTION
**Pull request by kubama**
## Description
Added proper authentication and authorization to snapshot endpoints

## Screenshots / Videos
N/A

## Checklist

- [X] My code follows the [UI Code Style](https://cloudifysource.atlassian.net/l/c/x8fq902p).
- [x] I followed the [UI Testing Guidelines](https://cloudifysource.atlassian.net/l/cp/udCVfvrx) and verified that all tests/checks have passed.
- [X] I verified if that change requires any update in the [documentation](https://cloudifysource.atlassian.net/l/c/4XKtPocy).
- [X] I added proper labels to this PR.

## Tests
Updated unit tests

 

## Documentation
N/A

### CODE REVIEW

File: backend/test/routes/snapshots/blueprint-layouts.test.ts [hunk 1]
```diff
@@ -3,6 +3,10 @@ import app from 'app';
 import mockDb from '../../mockDb';
 
 jest.mock('db/Connection');
+jest.mock('handler/AuthHandler', () => ({
+    isAuthorized: () => true,
+    getRBAC: () => Promise.resolve({ permissions: {} })
+}));
```
Comment (qooban) [inline backend/test/routes/snapshots/blueprint-layouts.test.ts:9]: Optional:

Is it possible to extract and share that mock across `snapshots/*.test.ts` files?

File: backend/routes/Snapshots.ts [hunk 2]
```diff
@@ -13,64 +15,91 @@ import { createUserWidgetsSnapshot, restoreUserWidgetsSnapshot } from '../handle
...
 
 const router = express.Router();
 
 router.use(express.json());
 
-function sendConflictResponse(res: Response) {
-    res.status(400).send({ message: 'Snapshot data conflicts with already existing data' });
+function getSnapshotPermissionValidator(permissionType: 'create' | 'restore'): RequestHandler<never, never> {
+    return async (req, res, next) => {
+        const rbac = await getRBAC(req.get('Authentication-Token')!);
```
Comment (qooban) [inline backend/routes/Snapshots.ts:26]: Optional:

In [backend/utils.ts#L17](https://github.com/cloudify-cosmo/cloudify-stage/blob/2e14efc088739bdf63e122aa63dff3198e0829f5/backend/utils.ts#L17) there is `getTokenFromCookies`. We could have there `getTokenFromHeader`.

File: backend/routes/Snapshots.ts [hunk 3]
```diff
@@ -13,64 +15,91 @@ import { createUserWidgetsSnapshot, restoreUserWidgetsSnapshot } from '../handle
...
+
+function defineSnapshotEndpoints<PayloadType>(
+    path: string,
+    createHandler: RequestHandler<never, PayloadType>,
+    restoreHandler: (req: Request<ParamsDictionary, any, PayloadType>) => Promise<unknown>
+) {
+    router.get(path, getSnapshotPermissionValidator('create'), createHandler);
+    router.post(
+        path,
+        getSnapshotPermissionValidator('restore'),
```
Comment (qooban) [inline backend/routes/Snapshots.ts:49]: I like that builder function and express middleware 👍🏻

**Review by qooban — DISMISSED** [review]
LGTM 👍🏻

File: backend/routes/Snapshots.ts (hunk 2 above)
Comment (kubama) [reply backend/routes/Snapshots.ts]: Done

**Review by PatrycjaKubiczek — APPROVED** [review]

**Review by qooban — APPROVED** [review]

**Pull request merged**